doc/man3/OSSL_SELF_TEST_new.pod

```diff
@@ -0,0 +1,177 @@
+=pod
+
+=head1 NAME
+
+OSSL_SELF_TEST_new,
+OSSL_SELF_TEST_free,
+OSSL_SELF_TEST_onbegin,
+OSSL_SELF_TEST_oncorrupt_byte,
+OSSL_SELF_TEST_onend - functionality to trigger a callback during a self test
+
+=head1 SYNOPSIS
+
+ #include <openssl/self_test.h>
+
+ OSSL_SELF_TEST *OSSL_SELF_TEST_new(OSSL_CALLBACK *cb, void *cbarg);
+ void OSSL_SELF_TEST_free(OSSL_SELF_TEST *st);
+
+ void OSSL_SELF_TEST_onbegin(OSSL_SELF_TEST *st, const char *type,
+                             const char *desc);
+ void OSSL_SELF_TEST_oncorrupt_byte(OSSL_SELF_TEST *st, unsigned char *bytes);
+ void OSSL_SELF_TEST_onend(OSSL_SELF_TEST *st, int ret);
+
+=head1 DESCRIPTION
+
+These methods are intended for use by provider implementors, to display
+diagnostic information during self testing.
+
+There are 3 types of self tests that can be set up:
+
+=over 4
+
+=item "Power on Self Tests" (POST)
+
+During startup a provider may want to run known answer tests on algorithms to
+test that the system is operating correctly.
+
+=item "On demand Self tests"
+
+The application may manually trigger the POST via an API call.
+
+=item "Conditional tests"
+
+Used during an operation on an algorithm e.g. when performing a pairwise test
+during key generation of asymmetric keys.
+
+=back
+
+OSSL_SELF_TEST_new() allocates an opaque B<OSSL_SELF_TEST> object that has a
+callback and callback argument associated with it. The callback is triggered by
+the 3 phases of the self test that can be setup.
+
+The following OSSL_PARAMS are passed to the callback during each phase:
```

Comment:
The description is specific to our FIPS provider, why was it moved from `doc/man7/OSSL_PROVIDER-FIPS.pod`?

Reply:
Because the pairwise consistency tests in keygen use these.
The keygen self tests can be activated in non fips mode now, so they are not just fips.
In a future PR, I will probably move most self test code out of the fips module so that it can be reused by other providers. If they use it, it will use the given events. If they dont like it they can write their own.

Reply:
Correction: Because *our implementation of* the pairwise consistency tests in keygen use these.

In other words, this is information about *our implementation* that is now inserted into the general reference documentation.  Other implementations could use whatever strings they want.

Reply:
Since this is information about *out implementation*, I would say it doesn't belong here.

Reply:
The 3 SELF_TEST_onXXX are visible here.. SO it does belong here.

Reply:
Argh...

Another implementation can use whatever strings they want, right?  They have no obligation whatsoever to use the strings you mention here in *their* calls to SELF_TEST_onXXX, do they?

I don't quite understand why it's so hard to understand that the description of an API and the description of *implementation specific strings* to that API are two different things?

Reply:
("implementation" here actually means "provider implementation")

Reply:
Will discuss with you tonight - you dont understand what I am saying either :)

Reply:
I have reworded it so hopefully we are on the same (man)page now. If not - tell me off tonight..

Reply:
numbers under 10 get written out as words, hence "three"

Reply:
Removed the docs containing the number 3 :)